akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed
+
+/**
+ * A behavior interceptor allows for intercepting message and signal reception and perform arbitrary logic -
+ * transform, filter, send to a side channel etc. It is the core API for decoration of behaviors. Many prewritten
+ * intercepting behaviors are provided through factories in the respective `Behaviors`.
+ *
+ * @tparam O The outer message type – the type of messages the intercepting behavior will accept
+ * @tparam I The inner message type - the type of message the wrapped behavior accepts
+ */
+// FIXME variance instead of crazy typecast-fest in the internals?
+abstract class BehaviorInterceptor[O, I] {
```

Comment:
might be worth adding this to both `scaladsl` and `javadsl`, for example exposing right `ActorContext`